source/agora/node/Ledger.d

```diff
@@ -356,6 +361,74 @@ public class Ledger
         return true;
     }
 
+    /***************************************************************************
+
+        Returns the public keys of the validators that signed the block at
+        the specified height, and and also not classified as 'missing' validator
+
+        Params:
+            height = block height
+
+        Returns: the public keys of the validators that signed the block at
+                 the specified height, and and also not
+                 classified as 'missing' validator
+
+    ***************************************************************************/
+
+    private Point[] getPKeysOfSignedNonMissingValidators (Height height) nothrow @safe
+    {
+        Point[] keys;
+        Block block;
+        try
+        {
+            block = this.storage.readBlock(height);
+        }
+        catch (Exception e) {
+            log.error("unable to read block at height: {}", height);
+            return null;
+        }
+
+        immutable validator_cnt = enroll_man.getCountOfValidators(height);
+        foreach (i; 0 .. validator_cnt)
+        {
+            auto validator_pkey = enroll_man.getValidatorAtIndex(height, i);
+            if (block.header.validators[i] && !block.header.missing_validators.canFind(i))
+                keys ~= validator_pkey;
+        }
+
+        return keys;
+    }
+
+    /***************************************************************************
+
+        Returns a map indexed by the public keys of validators who signed blocks
+            and also not classified as 'missing' validator at least once between
+            `start_height` and `end_height`
+
+        The value corresponding to the public key shows how many blocks the
+        validator signed between `start_height` and `end_height`
+
+        Params:
+            start_height = starting block height(inclusive)
+            end_height = ending block height(inclusive)
+
+        Returns: a map indexed by public keys of validators who signed blocks
+                 and also not classified as 'missing' validator at least once
+                 between `start_height` and `end_height`
+
+    ***************************************************************************/
+
+    private ushort[Point] getPKeysOfSignedNonMissingValidators (Height start_height, Height end_height) nothrow @safe
+    {
+        ushort[Point] signed_cnt_keys_map;
+
+        foreach (height; start_height .. end_height + 1)
+            foreach (point; getPKeysOfSignedNonMissingValidators(height))
+                signed_cnt_keys_map[point]++;
```

Comment:
Does this work ? I would expect it to throw.

Reply:
if you mean line 427, it works

```D
    int[string] m;
    m["hello"]++;
    writeln(m["hello"]);
```

writes out 1